crates/stages/src/stages/execution.rs

```diff
@@ -209,46 +209,67 @@ impl<DB: Database> Stage<DB> for ExecutionStage {
         for results in block_change_patches.into_iter() {
             // insert state change set
             for result in results.changesets.into_iter() {
-                // TODO insert to transitionId to tx_index
                 for (address, account_change_set) in result.changeset.into_iter() {
                     let AccountChangeSet { account, wipe_storage, storage } = account_change_set;
                     // apply account change to db. Updates AccountChangeSet and PlainAccountState
                     // tables.
                     trace!(target: "sync::stages::execution", ?address, current_transition_id, ?account, wipe_storage, "Applying account changeset");
                     account.apply_to_db(&**tx, address, current_transition_id)?;
 
-                    // wipe storage
+                    let storage_id = TransitionIdAddress((current_transition_id, address));
+
+                    // cast key to H256 and trace the change
+                    let storage = storage
+                        .into_iter()
+                        .map(|(key, (old_value,new_value))| {
+                            let hkey = H256(key.to_be_bytes());
+                            trace!(target: "sync::stages::execution", ?address, current_transition_id, ?hkey, ?old_value, ?new_value, "Applying storage changeset");
+                            (hkey, old_value,new_value)
+                        })
+                        .collect::<Vec<_>>();
+
+                    let mut cursor_storage_changeset =
+                        tx.cursor_write::<tables::StorageChangeSet>()?;
+
                     if wipe_storage {
-                        // TODO insert all changes to StorageChangeSet
+                        // iterate over storage and save them before entry is deleted.
+                        tx.cursor_read::<tables::PlainStorageState>()?
+                            .walk(address)?
+                            .take_while(|res| {
+                                res.as_ref().map(|(k, _)| *k == address).unwrap_or_default()
+                            })
+                            .try_for_each(|entry| {
+                                let (_, old_value) = entry?;
+                                cursor_storage_changeset.append(storage_id.clone(), old_value)
+                            })?;
+
+                        // delete all entries
```

Comment:
this is essentially:

> in case of selfdestruct read all storages of the account that got self-destructed and create changeset of all storage slots

right?

Reply:
Exactly!